assets/src/data/eventespresso/filter-state/index.js

```diff
@@ -0,0 +1,53 @@
+/**
+ * External dependencies
+ */
+import { Map } from 'immutable';
+import createSelector from 'rememo';
+import { registerStore } from '@wordpress/data';
+
+const DEFAULT_STATE = {};
+const SET_FILTER = 'SET_FILTER';
+
+const getState = ( state, context ) => {
+	return typeof state[ context ] === 'undefined' ?
+		Map() :
+		state[ context ];
+};
+
+const filterStateReducer = ( state = DEFAULT_STATE, action ) => {
+	const { type, context, filter, value } = action;
+	if ( type === SET_FILTER ) {
+		return {
+			...state,
+			[ context ]: getState( state, context ).set( filter, value ),
+		};
```

Comment:
This will return a new state object every time this specific action is dispatched even if there are no changes. This will result in unnecessary renders.

I realize that what I originally had in this pull was wrong too.  So probably what we need to do is something like:

```js
const filterStateReducer = ( state = DEFAULT_STATE, action ) => {
	const { type, context, filter, value } = action;
	if ( type === SET_FILTER ) {
		const newValue = getState( state, context ).set( filter, value );
		if ( state[ context ] !== newValue ) {
			return {
				...state,
				[ context ]: newValue,
			};
		}
	}
	return state;
};
```

As I did this, I realized that there's probably places in the existing store reducers where I'm  doing things wrong (mutating the original state) so I'll watch for that as a part of my review pull in #970. It's likely that there's a few places where I'm doing things incorrectly and that could be the cause of the wonky `withSelect` behaviour you were seeing.